lib/step.rb

```diff
@@ -137,32 +135,55 @@ def run(dry_run: false, fail_fast: false)
 
         return if passed?
 
-        puts @output
-        puts
-
-        @named_args.each do |name|
-          next if name.blank?
+        if ENV["GITHUB_ACTIONS"].blank?
+          puts @output
+        else
+          os_string = if OS.linux?
+            "Linux"
+          elsif Hardware::CPU.arm?
+            "macOS #{MacOS.version.pretty_name} (#{MacOS.version}) on Apple Silicon"
+          else
+            "macOS #{MacOS.version.pretty_name}"
+          end
 
-          path, line = begin
-            formula = Formulary.factory(name)
-            method_sym = command.second.to_sym
-            method_location = formula.method(method_sym).source_location if formula.respond_to?(method_sym)
+          @named_args.each do |name|
+            if name.blank?
+              puts @output
+              next
+            end
 
-            if method_location.present? && (method_location.first == formula.path.to_s)
-              method_location
-            else
-              [formula.path, nil]
+            path, line = begin
+              formula = Formulary.factory(name)
+              method_sym = command.second.to_sym
+              method_location = formula.method(method_sym).source_location if formula.respond_to?(method_sym)
+
+              if method_location.present? && (method_location.first == formula.path.to_s)
+                method_location
+              else
+                [formula.path, nil]
+              end
+            rescue FormulaUnavailableError
+              [@repository.glob("**/#{name}*").first, nil]
             end
-          rescue FormulaUnavailableError
-            [@repository.glob("**/#{name}*").first, nil]
-          end
-          next if path.blank?
+            if path.blank?
+              puts @output
+              next
+            end
+
+            annotation_type = failed? ? :error : :warning
+
+            # GitHub Actions has a 64KB maximum for annotiations. That's a bit
+            # too long so instead let's go for a maximum of 24KB or 256 lines.
+            max_length_start = [@output.length - (24 * 1024), 0].max
+            annotation_output = @output[max_length_start..].lines.last(256).join("\n")
```

Comment:
I have two concerns with this truncation scheme.

1. This will truncate output in the middle of a line. Not a huge deal, but we can probably fix this.
2. This means you never get to see the untruncated output in the log, which is often of interest (e.g. the output of `./configure`).

As a solution for 2, I suggest we still dump the full `@output` anyway, but hide it under a fold so we only open it up if we need to. https://docs.github.com/en/actions/learn-github-actions/workflow-commands-for-github-actions#grouping-log-lines

Reply:
@carlocab I'm punting on the middle-of-line truncation (unless we want to go with just line-based truncation which I'd be fine with).

I've added groups to output both types of output. Let me know what you think.

Reply:
I like this, but won't this only show truncated output in the CI log? I think we still want to see the full one too for this case.

Reply:
@carlocab Whoops, fixed.